Adds ability to copy status messages

initramfs opened the pull request:
This PR adds the ability to copy friends' status messages via a context menu placed upon the status message label within the chat form. The image below demonstrates the change:

![Context menu on status label](https://cloud.githubusercontent.com/assets/3128050/14765502/54cf4d7c-09b2-11e6-8d94-cab72aaa1ea7.png)

Closes issue #3155.

initramfs committed:
feat(status): add ability to copy status messages

Adds the ability to copy status messages via a context menu on the
status message label.

Closes issue #3155

zetok commented:
Status message shown in "About friend" window should also be copyable:
![about window](https://cloud.githubusercontent.com/assets/3148759/14765704/02a43c5a-09e5-11e6-82f5-ce299121d7e5.png)

initramfs committed:
feat(status): make status in about dialog copyable

initramfs commented:
@zetok Latest commit adds ability to copy status from "About friend" dialog

zetok commented:
Thanks :)

initramfs commented:
@zetok The build seems to be failing on travis for some weird reason, should that be a concern? A quick search on SO suggests the qmake version is out of date (on travis).

zetok commented:
> The build seems to be failing on travis for some weird reason, should that be a concern? A quick search on SO suggests the qmake version is out of date (on travis).

Yep. If failure isn't due to dependencies failing to download, then it's a genuine problem. And I doubt whether qmake is out-of-date, since for me it too fails to compile with Qt 5.5.1:

```
ccache g++ -c -O0 -ggdb -fstack-protector-strong -fno-exceptions -O2 -pthread -pthread -std=c++0x -Wall -W -D_REENTRANT -fPIC -DGIT_VERSION="\"07ff48badb79e0816dbc62566e005698e405764e\"" -DGIT_DESCRIBE="\"v1.3.0-251-g07ff48b\"" -DTIMESTAMP=1461479838 -DLOG_TO_FILE -DQT_MESSAGELOGCONTEXT -DQTOX_PLATFORM_EXT -DENABLE_SYSTRAY_STATUSNOTIFIER_BACKEND -DENABLE_SYSTRAY_GTK_BACKEND -DQT_NO_DEBUG -DQT_OPENGL_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_SQL_LIB -DQT_CORE_LIB -I. -Ilibs/include -isystem /usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -isystem /usr/include/gtk-2.0 -I/usr/lib64/gtk-2.0/include -isystem /usr/include/pango-1.0 -isystem /usr/include/atk-1.0 -isystem /usr/include/cairo -isystem /usr/include/pixman-1 -isystem /usr/include/libdrm -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/libpng16 -isystem /usr/include/harfbuzz -isystem /usr/include/freetype2 -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtOpenGL -isystem /usr/include/qt5/QtSvg -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtXml -isystem /usr/include/qt5/QtSql -isystem /usr/include/qt5/QtCore -I. -I. -I/usr/lib64/qt5/mkspecs/linux-g++ -o advancedform.o src/widget/form/settings/advancedform.cpp
In file included from /usr/include/qt5/QtCore/qiodevice.h:39:0,
                 from /usr/include/qt5/QtCore/qtextstream.h:37,
                 from /usr/include/qt5/QtCore/qdebug.h:42,
                 from /usr/include/qt5/QtCore/QDebug:1,
                 from src/widget/form/chatform.cpp:20:
/usr/include/qt5/QtCore/qobject.h: In lambda function:
/usr/include/qt5/QtCore/qobject.h:428:14: error: ‘QObject* QObject::sender() const’ is protected
     QObject *sender() const;
              ^
src/widget/form/chatform.cpp:122:69: error: within this context
             QWidget* sender = static_cast<QWidget*>(QObject::sender());
                                                                     ^
Makefile:3084: recipe for target 'chatform.o' failed
make: *** [chatform.o] Error 1
make: *** Waiting for unfinished jobs....
```

SkyzohKey commented:
Why are you limitating users with artificials fences ? Why not simply use a selectable label and let the user choose to select then copy only the part they want ?

initramfs commented:
@SkyzohKey I've implemented it in a way that emulates how most other chat clients do it. Besides, what happens if the message is too long?

As below:

![noselection](https://cloud.githubusercontent.com/assets/3128050/14791860/b543d374-0ae5-11e6-9d32-d6e6111ea44d.png)

A selectable label would need some way to account for overflow text, possibly involving scrolling which further complicates the UI.

Personally, I believe a simple copy + editing the message externally as necessary seems to be cleaner than implementing a lot of UI handling that complicates the way status messages work but I'm open to discussion if you disagree.

SkyzohKey commented:
@initramfs Isn't already the profile window displaying the whole status message ?
I never said that's a bad idea, but you can do both. x)

initramfs commented:
@SkyzohKey Yes it does, and frankly it's renders pretty badly on some DE/WMs

For example:
![baddialog](https://cloud.githubusercontent.com/assets/3128050/14792365/ea95fc62-0ae7-11e6-96f5-6ce81e09966d.png)

On my WM there is no way to resize that to be smaller than the full message length.

But I can see how a selection box there may be more appropriate. I guess I'll first implement something that breaks the status message into rows and then implement a selection mechanism for it.

SkyzohKey commented:
@initramfs That's the Qt magic, xD

zetok commented:
> I guess I'll first implement something that breaks the status message into rows and then implement a selection mechanism for it.

Well, `120 * \n + .`.

initramfs commented:
@zetok Lol no need, I'm doing it within the Qt form itself (it's basically done, just testing).

zetok commented:
@initramfs that's just a test case for it – i.e. how well it would handle statuses that span across dozens of lines. :)

initramfs committed:
refactor(status): enable text selection in about friend dialog

Enables text selection in the about friend dialog, allowing selection
and copying of status and username.

initramfs commented:
@zetok @SkyzohKey New changes have been implemented, example below:

![new](https://cloud.githubusercontent.com/assets/3128050/14794171/99abe5ba-0af0-11e6-8575-953f5b87c86f.png)

initramfs commented:
@zetok Regarding the travis CI compilation issue, what version of gcc/g++ are you using? I can't seem to reproduce the issue on Qt 5.6/5.3 and on Ubuntu trusty Qt 5.5.1. I know how to work around it if necessary...

zetok commented:
travis uses gcc 4.8, and I'm using 4.9.3

initramfs committed:
fix(status): explictly refer to this within lambda

Workaround for a potential gcc bug in which lambda expressions are not
being executed in the correct context.

initramfs commented:
@zetok I've fixed the compilation issue by working around it. Seems to be a GCC bug that affects the 4.x branch, potentially [this one](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61636). The latest commit contains a code change that's effectively the same as before but explictly uses `this` to "force" GCC to compile the lambda in the right context.
